SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc

```diff
@@ -1,461 +1,672 @@
+      previous bunches are added and only in-time, signal and after bunches
+      (defined by maximumSubsequentBunchCrossing_) are used.
+  */
+  const unsigned int maximumPreviousBunchCrossing_;
+  /** The maximum bunch crossing AFTER the signal crossing to create
+      TrackinParticles for. E.g. if set to zero only
+      uses the signal and in time pileup (and previous bunches defined by the
+      maximumPreviousBunchCrossing_ parameter).
+  */
+  const unsigned int maximumSubsequentBunchCrossing_;
+
+  const edm::InputTag simTrackLabel_;
+  const edm::InputTag simVertexLabel_;
+  edm::Handle<std::vector<SimTrack> > hSimTracks;
+  edm::Handle<std::vector<SimVertex> > hSimVertices;
+
+  std::vector<edm::InputTag> collectionTags_;
+  edm::InputTag genParticleLabel_;
+  /// Needed to add HepMC::GenVertex to SimVertex
+  edm::InputTag hepMCproductLabel_;
+
+  const double minEnergy_, maxPseudoRapidity_;
+
+ public:
+  struct OutputCollections {
+    std::unique_ptr<SimClusterCollection> pSimClusters;
+    std::unique_ptr<CaloParticleCollection> pCaloParticles;
+  };
+
+  struct calo_particles {
+    std::vector<uint32_t> sc_start_;
+    std::vector<uint32_t> sc_stop_;
+
+    void swap(calo_particles& oth) {
+      sc_start_.swap(oth.sc_start_);
+      sc_stop_.swap(oth.sc_stop_);
+    }
+
+    void clear() {
+      sc_start_.clear();
+      sc_stop_.clear();
+    }
+  };
+
+ private:
+  const HGCalTopology* hgtopo_[2];
+  const HGCalDDDConstants* hgddd_[2];
+  const HcalDDDRecConstants* hcddd_;
+  OutputCollections output_;
+  calo_particles m_caloParticles;
+};
+
+/* Graph utility functions */
+
+namespace {
+template <typename Edge, typename Graph, typename Visitor>
+void accumulateSimHits_edge(Edge& e, const Graph& g, Visitor* v) {
+  auto const edge_property = get(edge_weight, g, e);
+  v->total_simHits += edge_property.simHits;
+  IfLogDebug(DEBUG, messageCategory_)
+      << " Examining edges " << e << " --> particle " << edge_property.simTrack->type() << "("
+      << edge_property.simTrack->trackId() << ")"
+      << " with SimClusters: " << edge_property.simHits
+      << " Accumulated SimClusters: " << v->total_simHits << std::endl;
+}
+template <typename Vertex, typename Graph>
+void print_vertex(Vertex& u, const Graph& g) {
+  auto const vertex_property = get(vertex_name, g, u);
+  IfLogDebug(DEBUG, messageCategory_) << " At " << u;
+  // The Mother of all vertices has **no** SimTrack associated.
+  if (vertex_property.simTrack)
+    IfLogDebug(DEBUG, messageCategory_) << " [" << vertex_property.simTrack->type() << "]"
+                                        << "(" << vertex_property.simTrack->trackId() << ")";
+  IfLogDebug(DEBUG, messageCategory_) << std::endl;
+}
+
+// Graphviz output functions will only be generated in DEBUG mode
+#if DEBUG
+std::string graphviz_vertex(const VertexProperty& v) {
+  std::ostringstream oss;
+  oss << "{id: " << (v.simTrack ? v.simTrack->trackId() : 0)
```

Comment:
Maybe the graphviz output example would be a little easier to read if each property here were separated by newlines? (If that's supported in graphviz)

Reply:
> Maybe the graphviz output example would be a little easier to read if each property here were separated by newlines? (If that's supported in graphviz)

It does, apparently. Do you like it better?
[CaloParticleGraphExample.pdf](https://github.com/cms-sw/cmssw/files/1563155/CaloParticleGraphExample.pdf)

Reply:
Yes, much better!